jetty-webapp/src/main/java/org/eclipse/jetty/webapp/DecoratingListener.java

```diff
@@ -0,0 +1,178 @@
+//  ------------------------------------------------------------------------
+//  All rights reserved. This program and the accompanying materials
+//  are made available under the terms of the Eclipse Public License v1.0
+//  and Apache License v2.0 which accompanies this distribution.
+//
+//      The Eclipse Public License is available at
+//      http://www.eclipse.org/legal/epl-v10.html
+//
+//      The Apache License v2.0 is available at
+//      http://www.opensource.org/licenses/apache2.0.php
+//
+//  You may elect to redistribute this code under either of these licenses.
+//  ========================================================================
+//
+
+package org.eclipse.jetty.webapp;
+
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+import javax.servlet.ServletContextAttributeEvent;
+import javax.servlet.ServletContextAttributeListener;
+
+import org.eclipse.jetty.util.Decorator;
+import org.eclipse.jetty.util.log.Log;
+import org.eclipse.jetty.util.log.Logger;
+
+/**
+ * A ServletContextAttributeListener that listens for a specific attribute
+ * name (default "org.eclipse.jetty.webapp.Decorator") to obtain a
+ * decorator instance from the webapp.  The instance is then either coerced
+ * to a Decorator or reflected for decorator compatible methods so it can
+ * be added to the {@link WebAppContext#getObjectFactory()} as a
+ * {@link Decorator}.
+ */
+public class DecoratingListener implements ServletContextAttributeListener
+{
+    private static final Logger LOG = Log.getLogger(DecoratingListener.class);
+    private static final MethodType decorateType;
+    private static final MethodType destroyType;
+
+    static
+    {
+        try
+        {
+            MethodHandles.Lookup lookup = MethodHandles.lookup();
+            decorateType = MethodType.methodType(Object.class, Object.class);
+            destroyType = MethodType.methodType(Void.TYPE, Object.class);
+            // Ensure we have a match
+            lookup.findVirtual(Decorator.class, "decorate", decorateType);
+            lookup.findVirtual(Decorator.class, "destroy", destroyType);
+        }
+        catch (Exception e)
+        {
+            throw new IllegalStateException(e);
+        }
+    }
+
+    private final WebAppContext _context;
+    private final String _attributeName;
+    private Decorator _decorator;
+
+    public DecoratingListener()
+    {
+        this(null, null);
+    }
+
+    public DecoratingListener(String attributeName)
+    {
+        this(null, attributeName);
+    }
+
+    public DecoratingListener(WebAppContext context)
+    {
+        this(context, null);
+    }
+
+    public DecoratingListener(WebAppContext context, String attributeName)
+    {
+        _context = context == null ? WebAppContext.getCurrentWebAppContext() : context;
+        _attributeName = attributeName == null ? "org.eclipse.jetty.webapp.Decorator" : attributeName;
```

Comment:
I have doubts on the attribute name, as it looks like a class, but the actual `Decorator` class is in a different package.
I would use `org.eclipse.jetty.webapp.decorator` instead.
Also put it somewhere as a constant.

Reply:
Good point.  The Decorator is part of the `org.eclipse.jetty.util` package currently.
It's not a webapp specific concept.
Client apps can use it too. (like a standalone weld instance does).